src/main/kotlin/com/cognifide/gradle/aem/pkg/DownloadTask.kt

```diff
@@ -0,0 +1,141 @@
+
+import com.cognifide.gradle.aem.api.AemConfig
+import com.cognifide.gradle.aem.api.AemTask
+import com.cognifide.gradle.aem.base.vlt.CheckoutConfig
+import com.cognifide.gradle.aem.instance.Instance
+import com.cognifide.gradle.aem.instance.InstanceSync
+import com.cognifide.gradle.aem.internal.PropertyParser
+import com.cognifide.gradle.aem.internal.file.FileContentReader
+import com.cognifide.gradle.aem.internal.file.FileException
+import org.gradle.api.file.DuplicatesStrategy
+import org.gradle.api.tasks.*
+import org.gradle.api.tasks.bundling.Zip
+import java.io.File
+
+open class DownloadTask : Zip(), AemTask {
+
+    companion object {
+        const val NAME = "aemDownload"
+        const val CLASSIFIER_DOWNLOAD = "download"
+        const val CLASSIFIER_SHELL = "shell"
+        const val EXTRACT_FLAG = "aem.download.extract"
+        const val FORCE_NEW_FLAG = "aem.download.extract.force.new"
+    }
+
+    @Nested
+    final override val config = AemConfig.of(project)
+
+    @InputDirectory
+    private val prepareDir = AemTask.taskDir(project, PrepareTask.NAME)
+
+    @Internal
+    val props = PropertyParser(project)
+
+    @Input
+    val forceNewFlag = props.flag(FORCE_NEW_FLAG)
+
+    @Input
+    val extractFlag = props.flag(EXTRACT_FLAG)
+
+    private val checkoutConfig = CheckoutConfig(project, config)
+
+    @Input
+    var filterPathProp: String = props.string("aem.download.filterPath", "")
+
+    @Input
+    var filterRootsProp = props.list("aem.download.filterRoots")
+
+    private val checkoutFilter by lazy { checkoutConfig.determineFilter(filterRootsProp, filterPathProp) }
+
+    @Input
+    val instanceProp = props.string("aem.download.instance", "")
+
+    private val instance: Instance by lazy { checkoutConfig.determineInstance(instanceProp) }
+
+    init {
+        description = "Builds and downloads CRX package from remote instance"
+        group = AemTask.GROUP
+
+        baseName = config.packageName
+        isZip64 = true
+
+        //Empty package uploaded to aem will have 'shell' suffix
+        classifier = CLASSIFIER_SHELL
+
+        //Take only first filter.xml file from definition
+        duplicatesStrategy = DuplicatesStrategy.EXCLUDE
+
+        //Load zip configuration only when the task is in the graph
+        project.gradle.taskGraph.whenReady {
+            if (it.hasTask(this)) {
+                this.specShellPackage()
+            }
+        }
+
+        // Task is always executed when in the graph
+        outputs.upToDateWhen { false }
+    }
+
+    @TaskAction
+    override fun copy() {
+        //Prepare shell package
+        super.copy()
+
+        val sync = InstanceSync(project, instance)
+        logger.lifecycle("Uploading package $archivePath")
+        val packagePath = sync.uploadPackage(archivePath).path
+        logger.lifecycle("Building remote package $packagePath")
+        sync.buildPackage(packagePath)
+        logger.lifecycle("Downloading remote package $packagePath")
```

Comment:
in my opinion you should change  `downloadPackage(packagePath, destinationDir)`  ->  `downloadPackage(packagePath, packageFile)` , I mean second param should not be a directory but rather a file, your contract is more constrained / less flexible. if you do that we could write more detailed log entry:

![image](https://user-images.githubusercontent.com/10673561/45572613-08e3ef80-b86a-11e8-8168-59946175b287.png)